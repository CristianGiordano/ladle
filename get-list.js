const { parse } = require("@babel/parser");
const traverse = require("@babel/traverse").default;
const template = require("@babel/template").default;
const generate = require("@babel/generator").default;
const t = require("@babel/types");
const fs = require("fs").promises;
const path = require("path");
const { relAppPath } = require("./const");

const plugins = [
  "jsx",
  "asyncGenerators",
  "classProperties",
  "classPrivateProperties",
  "classPrivateMethods",
  [
    "decorators",
    {
      decoratorsBeforeExport: true,
    },
  ],
  "doExpressions",
  "dynamicImport",
  "exportDefaultFrom",
  "exportNamespaceFrom",
  "functionBind",
  "functionSent",
  "importMeta",
  "logicalAssignment",
  "nullishCoalescingOperator",
  "numericSeparator",
  "objectRestSpread",
  "optionalCatchBinding",
  "optionalChaining",
  "partialApplication",
  "throwExpressions",
  "topLevelAwait",
];

const capitalize = (s) => {
  if (typeof s !== "string") return "";
  return s.charAt(0).toUpperCase() + s.slice(1);
};

const getStories = (stories) => {
  return generate(
    t.exportNamedDeclaration(
      t.variableDeclaration("const", [
        t.variableDeclarator(
          t.identifier("stories"),
          t.objectExpression(
            stories.map((story) =>
              t.objectProperty(
                t.stringLiteral(story),
                t.objectExpression([
                  t.objectProperty(
                    t.identifier("component"),
                    t.identifier(
                      `${story.split("--")[0]}${
                        story.split("--")[1]
                      }`.toLowerCase()
                    )
                  ),
                  t.objectProperty(
                    t.identifier("name"),
                    t.stringLiteral(capitalize(story.split("--")[1]))
                  ),
                ])
              )
            )
          )
        ),
      ])
    )
  ).code;
};

const getList = async (entries) => {
  let output = `import { lazy } from "react";\n`;
  const lazyImport = template(`
    const %%component%% = lazy(() =>
     import(%%source%%).then((module) => {
        return { default: module.%%story%% };
      })
    );
  `);
  const stories = [];
  for (entry of entries) {
    const fileId = entry.split("/")[1].split(".")[0];
    const code = await fs.readFile(path.join("./", entry), "utf8");
    const ast = parse(code, {
      sourceType: "module",
      plugins: [
        ...plugins,
        entry.endsWith(".ts") || entry.endsWith(".tsx") ? "typescript" : "flow",
      ],
    });
    traverse(ast, {
      ExportNamedDeclaration: (astPath) => {
        const storyName = astPath.node.declaration.declarations[0].id.name;
        const storyId = `${fileId}--${storyName.toLowerCase()}`;
        stories.push(storyId);
        const ast = lazyImport({
          source: t.stringLiteral(path.join(relAppPath, entry)),
          component: t.identifier(`${fileId}${storyName}`.toLocaleLowerCase()),
          story: t.identifier(storyName),
        });
        output += `\n${generate(ast).code}`;
      },
    });
  }
  output += `\n\n${getStories(stories)}`;
  return output;
};

module.exports = getList;
