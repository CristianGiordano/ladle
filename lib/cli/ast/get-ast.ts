import { parse, ParserPlugin } from "@babel/parser";

const plugins: ParserPlugin[] = [
  "jsx",
  "asyncGenerators",
  "classProperties",
  "classPrivateProperties",
  "classPrivateMethods",
  [
    "decorators",
    {
      decoratorsBeforeExport: true,
    },
  ],
  "doExpressions",
  "dynamicImport",
  "exportDefaultFrom",
  "exportNamespaceFrom",
  "functionBind",
  "functionSent",
  "importMeta",
  "logicalAssignment",
  "nullishCoalescingOperator",
  "numericSeparator",
  "objectRestSpread",
  "optionalCatchBinding",
  "optionalChaining",
  "partialApplication",
  "throwExpressions",
  "topLevelAwait",
];

const getAst = (code: string, filename: string) =>
  parse(code, {
    sourceType: "module",
    plugins: [
      ...plugins,
      filename.endsWith(".ts") || filename.endsWith(".tsx")
        ? "typescript"
        : "flow",
    ],
  });

export default getAst;
