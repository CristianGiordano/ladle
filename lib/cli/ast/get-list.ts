import traverse from "@babel/traverse";
import template from "@babel/template";
import generate from "@babel/generator";
import * as t from "@babel/types";
import { promises as fs } from "fs";
import path from "path";
import {
  getFileId,
  getEncodedStoryName,
  storyDelimiter,
  storyEncodeDelimiter,
  titleToFileId,
} from "../../app/src/story-name";
import { appSrcDir } from "../const";
import { kebabCase } from "../utils";
import getAst from "./get-ast";
import { converter } from "./ast-to-obj";

const getStories = (stories: string[]) => {
  return generate(
    t.exportNamedDeclaration(
      t.variableDeclaration("let", [
        t.variableDeclarator(
          t.identifier("stories"),
          t.objectExpression(
            stories.map((story) =>
              t.objectProperty(
                t.stringLiteral(story),
                t.objectExpression([
                  t.objectProperty(
                    t.identifier("component"),
                    t.identifier(
                      story.replace(
                        new RegExp(storyDelimiter, "g"),
                        storyEncodeDelimiter
                      )
                    )
                  ),
                ])
              )
            )
          )
        ),
      ])
    )
  ).code;
};

export const getList = async (entries: string[]) => {
  let output = `import { lazy } from "react";\n`;
  const lazyImport = template(`
    const %%component%% = lazy(() =>
     import(%%source%%).then((module) => {
        return { default: module.%%story%% };
      })
    );
  `);
  const stories: string[] = [];
  for (let entry of entries) {
    let exportDefaultProps: any = null;
    let exportDefault: any = null;
    const namedExportToStoryName: { [key: string]: string } = {};
    const fileId = getFileId(entry);
    const code = await fs.readFile(path.join("./", entry), "utf8");
    const ast = getAst(code, entry);
    traverse(ast, {
      Program(astPath: any) {
        astPath.node.body.forEach((child: any) => {
          if (
            child.type === "ExpressionStatement" &&
            child.expression.left &&
            child.expression.left.property &&
            child.expression.left.property.name === "storyName"
          ) {
            const storyExport = child.expression.left.object.name;
            if (child.expression.right.type !== "StringLiteral") {
              console.log(`${storyExport}.storyName must be a string literal.`);
            } else {
              namedExportToStoryName[storyExport] =
                child.expression.right.value;
            }
          }
        });
      },
      ExportDefaultDeclaration(astPath: any) {
        if (!astPath) return;
        exportDefault = astPath.node.declaration;
        try {
          const obj = converter(exportDefault);
          const json = JSON.stringify(obj);
          exportDefaultProps = JSON.parse(json);
        } catch (e) {
          console.warn("Default export parsing failed.");
        }
      },
      ExportNamedDeclaration: (astPath: any) => {
        const namedExport: string =
          astPath.node.declaration.declarations[0].id.name;
        let storyNamespace = fileId;
        if (exportDefaultProps && exportDefaultProps.title) {
          storyNamespace = titleToFileId(exportDefaultProps.title);
        }
        const storyName = namedExportToStoryName[namedExport]
          ? namedExportToStoryName[namedExport]
          : namedExport;
        const storyId = `${kebabCase(
          storyNamespace
        )}${storyDelimiter}${storyDelimiter}${kebabCase(storyName)}`;
        stories.push(storyId);
        const componentName = getEncodedStoryName(
          kebabCase(storyNamespace),
          kebabCase(storyName)
        );
        const ast = lazyImport({
          source: t.stringLiteral(
            path.join(path.relative(appSrcDir, process.cwd()), entry)
          ),
          component: t.identifier(componentName),
          story: t.identifier(namedExport),
        });
        output += `\n${generate(ast as any).code}`;
      },
    });
  }
  return `${output}\n\n${getStories(stories)}`;
};

export const getListWithHmr = async (entries: string[]) => {
  const hot = `if (import.meta.hot) {
    import.meta.hot.accept(({ module }) => {
      if (Object.keys(module.stories).every(item => Object.keys(stories).includes(item))) {
        stories = module.stories;
      } else {
        // full refresh when new stories are added
        // todo, can dynamic import + React Refresh work?
        import.meta.hot.invalidate();
      }
    });
  }`;
  return `${await getList(entries)}\n${hot}`;
};
