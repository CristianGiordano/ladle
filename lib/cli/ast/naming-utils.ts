export const storyDelimiter = "-";
export const storyEncodeDelimiter = "$";

export const titleToFileId = (title: string) =>
  title
    .toLocaleLowerCase()
    .replace(/\s*\/\s*/g, `${storyDelimiter}${storyDelimiter}`)
    .replace(/\s+/g, storyDelimiter);

export const getFileId = (filename: string) => {
  const pathParts = filename.split("/");
  return pathParts[pathParts.length - 1].split(".")[0];
};

export const getEncodedStoryName = (fileId: string, namedExport: string) => {
  return `${fileId}${storyEncodeDelimiter}${storyEncodeDelimiter}${namedExport}`
    .toLocaleLowerCase()
    .replace(new RegExp(storyDelimiter, "g"), storyEncodeDelimiter);
};
