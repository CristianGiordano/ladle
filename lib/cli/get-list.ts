import { parse, ParserPlugin } from "@babel/parser";
import traverse from "@babel/traverse";
import template from "@babel/template";
import generate from "@babel/generator";
import * as t from "@babel/types";
import { promises as fs } from "fs";
import path from "path";
import {
  getFileId,
  getEncodedStoryName,
  storyDelimiter,
  storyEncodeDelimiter,
  //capitalize,
} from "../app/src/story-name";
import { kebabCase } from "./utils";

const plugins: ParserPlugin[] = [
  "jsx",
  "asyncGenerators",
  "classProperties",
  "classPrivateProperties",
  "classPrivateMethods",
  [
    "decorators",
    {
      decoratorsBeforeExport: true,
    },
  ],
  "doExpressions",
  "dynamicImport",
  "exportDefaultFrom",
  "exportNamespaceFrom",
  "functionBind",
  "functionSent",
  "importMeta",
  "logicalAssignment",
  "nullishCoalescingOperator",
  "numericSeparator",
  "objectRestSpread",
  "optionalCatchBinding",
  "optionalChaining",
  "partialApplication",
  "throwExpressions",
  "topLevelAwait",
];

const getStories = (stories: string[]) => {
  return generate(
    t.exportNamedDeclaration(
      t.variableDeclaration("let", [
        t.variableDeclarator(
          t.identifier("stories"),
          t.objectExpression(
            stories.map((story) =>
              t.objectProperty(
                t.stringLiteral(story),
                t.objectExpression([
                  t.objectProperty(
                    t.identifier("component"),
                    t.identifier(
                      story.replace(
                        new RegExp(storyDelimiter, "g"),
                        storyEncodeDelimiter
                      )
                    )
                  ),
                ])
              )
            )
          )
        ),
      ])
    )
  ).code;
};

const getList = async (entries: string[], cacheDir: string) => {
  let output = `import { lazy } from "react";\n`;
  const lazyImport = template(`
    export var %%component%% = lazy(() =>
     import(%%source%%).then((module) => {
        return { default: module.%%story%% };
      })
    );
  `);
  const stories: string[] = [];
  for (let entry of entries) {
    const fileId = getFileId(entry);
    const code = await fs.readFile(path.join("./", entry), "utf8");
    const ast = parse(code, {
      sourceType: "module",
      plugins: [
        ...plugins,
        entry.endsWith(".ts") || entry.endsWith(".tsx") ? "typescript" : "flow",
      ],
    });
    traverse(ast, {
      ExportNamedDeclaration: (astPath: any) => {
        const storyName = astPath.node.declaration.declarations[0].id.name;
        const storyId = `${kebabCase(
          fileId
        )}${storyDelimiter}${storyDelimiter}${kebabCase(storyName)}`;
        stories.push(storyId);
        const componentName = getEncodedStoryName(
          kebabCase(fileId),
          kebabCase(storyName)
        );
        const ast = lazyImport({
          source: t.stringLiteral(
            path.join(path.relative(path.join(cacheDir), process.cwd()), entry)
          ),
          component: t.identifier(componentName),
          story: t.identifier(storyName),
        });
        output += `\n${generate(ast as any).code}`;
      },
    });
  }
  const hot = `if (import.meta.hot) {
    import.meta.hot.accept(({ module }) => {
      if (module.list.every(item => list.includes(item))) {
        stories = module.stories;
        list = module.list;
      } else {
        // full refresh when new stories are added
        // todo, can dynamic import + React Refresh work?
        import.meta.hot.invalidate();
      }
    });
  }`;
  output += `\n\n${getStories(stories)}`;
  return `${output}\nexport let list = ${JSON.stringify(stories)}\n${hot}`;
};

export default getList;
