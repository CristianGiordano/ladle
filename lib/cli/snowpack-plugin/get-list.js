const traverse = require("@babel/traverse").default;
const template = require("@babel/template").default;
const generate = require("@babel/generator").default;
const t = require("@babel/types");
const fs = require("fs");
const path = require("path");
const clonedeep = require("lodash.clonedeep");
const merge = require("lodash.merge");
const {
  getFileId,
  getEncodedStoryName,
  storyDelimiter,
  storyEncodeDelimiter,
  titleToFileId,
  kebabCase,
} = require("./naming-utils.js");
const { appSrcDir } = require("./const.js");
const getAst = require("./get-ast.js");
const { converter } = require("./ast-to-obj.js");

/**
 * @param {string[]} stories
 * @param {any} storiesParams
 */
const getStories = (stories, storiesParams) => {
  return generate(
    t.exportNamedDeclaration(
      t.variableDeclaration("let", [
        t.variableDeclarator(
          t.identifier("stories"),
          t.objectExpression(
            stories.map((story) => {
              let paramsAst = null;
              if (storiesParams[story]) {
                paramsAst = t.objectProperty(
                  t.identifier("parameters"),
                  /** @type {any} */ (template.ast(
                    `const foo = ${JSON.stringify(storiesParams[story])}`
                  )).declarations[0].init
                );
              }
              return t.objectProperty(
                t.stringLiteral(story),
                t.objectExpression([
                  t.objectProperty(
                    t.identifier("component"),
                    t.identifier(
                      story.replace(
                        new RegExp(storyDelimiter, "g"),
                        storyEncodeDelimiter
                      )
                    )
                  ),
                  ...(paramsAst ? [paramsAst] : []),
                ])
              );
            })
          )
        ),
      ])
    )
  ).code;
};

/**
 * @param {string[]} entries
 */
const getList = async (entries) => {
  let output = `import { lazy } from "react";\n`;
  const lazyImport = template(`
    const %%component%% = lazy(() =>
     import(%%source%%).then((module) => {
        return { default: module.%%story%% };
      })
    );
  `);
  /**
   * @type {string []}
   */
  const stories = [];
  /**
   * @type {any}
   */
  const storiesParams = {};
  for (let entry of entries) {
    /**
     * @type {any}
     */
    let exportDefaultProps = null;
    /**
     * @type {{[key: string]: string}}
     */
    const namedExportToStoryName = {};
    /**
     * @type {{[key: string]: any}}
     */
    const namedExportToParameters = {};
    const fileId = getFileId(entry);
    const code = await fs.promises.readFile(path.join("./", entry), "utf8");
    const ast = getAst(code, entry);
    traverse(ast, {
      // extracting Story.storyName and Story.parameters
      /**
       * @param {any} astPath
       */
      Program(astPath) {
        astPath.node.body.forEach((/** @type {any} */ child) => {
          if (
            child.type === "ExpressionStatement" &&
            child.expression.left &&
            child.expression.left.property
          ) {
            if (child.expression.left.property.name === "storyName") {
              const storyExport = child.expression.left.object.name;
              if (child.expression.right.type !== "StringLiteral") {
                console.log(
                  `${storyExport}.storyName must be a string literal.`
                );
              } else {
                namedExportToStoryName[storyExport] =
                  child.expression.right.value;
              }
            } else if (child.expression.left.property.name === "parameters") {
              const storyExport = child.expression.left.object.name;
              if (child.expression.right.type !== "ObjectExpression") {
                console.log(
                  `${storyExport}.parameters must be an object expression.`
                );
              } else {
                const obj = converter(child.expression.right);
                const json = JSON.stringify(obj);
                namedExportToParameters[storyExport] = JSON.parse(json);
              }
            }
          }
        });
      },
      // extracting default export object
      /**
       * @param {any} astPath
       */
      ExportDefaultDeclaration(astPath) {
        if (!astPath) return;
        try {
          const obj = converter(astPath.node.declaration);
          const json = JSON.stringify(obj);
          exportDefaultProps = JSON.parse(json);
        } catch (e) {
          console.warn("Default export parsing failed.");
        }
      },
      // extracting story names (named exports)
      /**
       * @param {any} astPath
       */
      ExportNamedDeclaration: (astPath) => {
        /**
         * @type {string}
         */
        const namedExport = astPath.node.declaration.declarations[0].id.name;
        let storyNamespace = fileId;
        if (exportDefaultProps && exportDefaultProps.title) {
          storyNamespace = titleToFileId(exportDefaultProps.title);
        }
        const storyName = namedExportToStoryName[namedExport]
          ? namedExportToStoryName[namedExport]
          : namedExport;
        const storyId = `${kebabCase(
          storyNamespace
        )}${storyDelimiter}${storyDelimiter}${kebabCase(storyName)}`;
        stories.push(storyId);
        // attach default parameters to each story
        if (exportDefaultProps && exportDefaultProps.parameters) {
          storiesParams[storyId] = exportDefaultProps;
        }
        // add and merge story specific parameters
        if (namedExportToParameters[namedExport]) {
          storiesParams[storyId] = merge(
            clonedeep(storiesParams[storyId] || {}),
            {
              parameters: namedExportToParameters[namedExport],
            }
          );
        }

        const componentName = getEncodedStoryName(
          kebabCase(storyNamespace),
          kebabCase(storyName)
        );
        const ast = lazyImport({
          source: t.stringLiteral(
            path.join(path.relative(appSrcDir, process.cwd()), entry)
          ),
          component: t.identifier(componentName),
          story: t.identifier(namedExport),
        });
        output += `\n${generate(/** @type {any} */ (ast)).code}`;
      },
    });
  }
  return `${output}\n\n${getStories(stories, storiesParams)}`;
};

const getConfigImport = () => {
  const configPath = path.join(process.cwd(), "./.ladle/config.mjs");
  const configExists = fs.existsSync(configPath);
  const relativePath = path.relative(
    path.join(__dirname, "../../app/src"),
    configPath
  );
  let configCode = `export let config = {};\n`;
  if (configExists) {
    configCode += `import customConfig from '${relativePath}';\nconfig = customConfig;\n`;
  }
  return `${configCode}`;
};

/**
 * @param {string[]} entries
 */
const getListWithHmr = async (entries) => {
  const hot = `if (import.meta.hot) {
    import.meta.hot.accept(({ module }) => {
      if (Object.keys(module.stories).every(item => Object.keys(stories).includes(item))) {
        stories = module.stories;
        config = module.config;
      } else {
        // full refresh when new stories are added
        // todo, can dynamic import + React Refresh work?
        import.meta.hot.invalidate();
      }
    });
  }`;
  return `${await getList(entries)}\n${getConfigImport()}\n${hot}`;
};

module.exports = {
  getList,
  getListWithHmr,
};
