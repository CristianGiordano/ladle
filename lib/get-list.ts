import { parse, ParserPlugin } from "@babel/parser";
import traverse from "@babel/traverse";
import template from "@babel/template";
import generate from "@babel/generator";
import * as t from "@babel/types";
import { promises as fs } from "fs";
import path from "path";
import {
  getFileId,
  getEncodedStoryName,
  storyDelimiter,
  storyEncodeDelimiter,
} from "./app/story-name";

const plugins: ParserPlugin[] = [
  "jsx",
  "asyncGenerators",
  "classProperties",
  "classPrivateProperties",
  "classPrivateMethods",
  [
    "decorators",
    {
      decoratorsBeforeExport: true,
    },
  ],
  "doExpressions",
  "dynamicImport",
  "exportDefaultFrom",
  "exportNamespaceFrom",
  "functionBind",
  "functionSent",
  "importMeta",
  "logicalAssignment",
  "nullishCoalescingOperator",
  "numericSeparator",
  "objectRestSpread",
  "optionalCatchBinding",
  "optionalChaining",
  "partialApplication",
  "throwExpressions",
  "topLevelAwait",
];

const getStories = (stories: string[]) => {
  return generate(
    t.exportNamedDeclaration(
      t.variableDeclaration("const", [
        t.variableDeclarator(
          t.identifier("stories"),
          t.objectExpression(
            stories.map((story) =>
              t.objectProperty(
                t.stringLiteral(story),
                t.objectExpression([
                  t.objectProperty(
                    t.identifier("component"),
                    t.identifier(
                      story.replace(
                        new RegExp(storyDelimiter, "g"),
                        storyEncodeDelimiter
                      )
                    )
                  ),
                ])
              )
            )
          )
        ),
      ])
    )
  ).code;
};

const getList = async (entries: string[], cacheDir: string) => {
  let output = `import { lazy } from "react";\n`;
  const lazyImport = template(`
    const %%component%% = lazy(() =>
     import(%%source%%).then((module) => {
        return { default: module.%%story%% };
      })
    );
  `);
  const stories: string[] = [];
  for (let entry of entries) {
    const fileId = getFileId(entry);
    const code = await fs.readFile(path.join("./", entry), "utf8");
    const ast = parse(code, {
      sourceType: "module",
      plugins: [
        ...plugins,
        entry.endsWith(".ts") || entry.endsWith(".tsx") ? "typescript" : "flow",
      ],
    });
    traverse(ast, {
      ExportNamedDeclaration: (astPath: any) => {
        const storyName = astPath.node.declaration.declarations[0].id.name;
        const storyId = `${fileId}${storyDelimiter}${storyDelimiter}${storyName.toLowerCase()}`;
        stories.push(storyId);
        const ast = lazyImport({
          source: t.stringLiteral(
            path.join(
              path.relative(path.join(cacheDir, "app"), process.cwd()),
              entry
            )
          ),
          component: t.identifier(getEncodedStoryName(fileId, storyName)),
          story: t.identifier(storyName),
        });
        output += `\n${generate(ast as any).code}`;
      },
    });
  }
  output += `\n\n${getStories(stories)}`;
  return output;
};

export default getList;
