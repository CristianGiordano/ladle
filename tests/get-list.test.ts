import { getList } from "../lib/cli/ast/get-list";

test("Single file with two stories", async () => {
  expect(
    await getList(["./tests/fixtures/animals.stories.tsx"])
  ).toMatchSnapshot();
});

test("Capital letters in story names converted into delimiters", async () => {
  expect(
    await getList(["./tests/fixtures/capitalization.stories.tsx"])
  ).toMatchSnapshot();
});

test("Capital letters in the filename converted into delimiters", async () => {
  expect(
    await getList(["./tests/fixtures/filenameCapitalization.stories.tsx"])
  ).toMatchSnapshot();
});

test("Turn file name delimiters into spaces and levels correctly", async () => {
  expect(
    await getList(["./tests/fixtures/our-animals--mammals.stories.tsx"])
  ).toMatchSnapshot();
});

test("Default title is used instead of the file name", async () => {
  expect(
    await getList(["./tests/fixtures/default-title.stories.tsx"])
  ).toMatchSnapshot();
});

test("Story name replaces named export as a story name", async () => {
  expect(
    await getList(["./tests/fixtures/storyname.stories.tsx"])
  ).toMatchSnapshot();
});

test("Extract default parameters", async () => {
  expect(
    await getList(["./tests/fixtures/default-parameters.stories.tsx"])
  ).toMatchSnapshot();
});
