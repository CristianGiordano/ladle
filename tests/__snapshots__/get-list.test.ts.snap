// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Capital letters in story names converted into delimiters 1`] = `
"import { lazy } from \\"react\\";

const capitalization$$blue$tiny$cat = lazy(() => import(\\"../../tests/fixtures/capitalization.stories.tsx\\").then(module => {
  return {
    default: module.BlueTinyCat
  };
}));
const capitalization$$big$barking$dog = lazy(() => import(\\"../../tests/fixtures/capitalization.stories.tsx\\").then(module => {
  return {
    default: module.BigBarkingDog
  };
}));

export const stories = {
  \\"capitalization--blue-tiny-cat\\": {
    component: capitalization$$blue$tiny$cat
  },
  \\"capitalization--big-barking-dog\\": {
    component: capitalization$$big$barking$dog
  }
};"
`;

exports[`Capital letters in the filename converted into delimiters 1`] = `
"import { lazy } from \\"react\\";

const filename$capitalization$$test = lazy(() => import(\\"../../tests/fixtures/filenameCapitalization.stories.tsx\\").then(module => {
  return {
    default: module.Test
  };
}));

export const stories = {
  \\"filename-capitalization--test\\": {
    component: filename$capitalization$$test
  }
};"
`;

exports[`Single file with two stories 1`] = `
"import { lazy } from \\"react\\";

const animals$$cat = lazy(() => import(\\"../../tests/fixtures/animals.stories.tsx\\").then(module => {
  return {
    default: module.Cat
  };
}));
const animals$$dog = lazy(() => import(\\"../../tests/fixtures/animals.stories.tsx\\").then(module => {
  return {
    default: module.Dog
  };
}));

export const stories = {
  \\"animals--cat\\": {
    component: animals$$cat
  },
  \\"animals--dog\\": {
    component: animals$$dog
  }
};"
`;

exports[`Turn file name delimiters into spaces and levels correctly 1`] = `
"import { lazy } from \\"react\\";

const our$animals$$mammals$$cat = lazy(() => import(\\"../../tests/fixtures/our-animals--mammals.stories.tsx\\").then(module => {
  return {
    default: module.Cat
  };
}));

export const stories = {
  \\"our-animals--mammals--cat\\": {
    component: our$animals$$mammals$$cat
  }
};"
`;
